README.md
# awesome_ui_button

![](https://github.com/Flutter-Awesome-Libs/flutter_awesome_button/workflows/build/badge.svg)
[![License: MIT](https://img.shields.io/badge/License-MIT-brightgreen.svg)](https://github.com/Flutter-Awesome-Libs/flutter_awesome_button/blob/master/LICENSE)

This package allows you to add a good looking button to your application easily and without having to write a lot of code yourself, simply import and start using.

## Add the dependency

Add this to your pubspec.yaml
```
dependencies:
  awesome_ui_button: ^1.0.3
```

## Installing

Run the following command

```
flutter pub get
```

## Import the package

```
import 'package:awesome_ui_button/awesome_ui_button.dart';
```

## Add the button to your application

```dart
AwesomeButton(
  text: "Default",
),

OR

AwesomeButton.outline(
  text: "Outline",
),


OR

AwesomeButton.destructive(
  text: "Destructive",
),
```

For full documentation visit [flutterawesome.dev](https://flutterawesome.dev/docs/components/button)


CHANGELOG.md
## [1.0.1 + 1.0.2 + 1.0.3] - Aug 2020

* Fixes an issue where colors would not reset correctly when disabled is toggled from true to false
* Honors key being passed in constructor
* Fixes issue with tap behavior

## [1.0.0] - Jan 2020

* Release of 1.0.0

## [0.0.1] - Jan 2020

* Internal release(do not use in production)

test/default_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'dart:math' show max;

import 'package:awesome_ui_button/awesome_ui_button.dart';

void main() {
  Color _getActiveColor(Color baseColor) {
    int red = max(baseColor.red - 20, 0);
    int green = max(baseColor.green - 40, 0);
    int blue = max(baseColor.blue - 40, 0);

    return Color.fromRGBO(red, green, blue, 1);
  }

  /**
   * MARK: Test that the UI renders correctly for all situations
   */

  testWidgets("Test that default button renders correctly", (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(text: text),
        ),
      ),
    );

    Finder gestureFinder = find.byType(GestureDetector);
    Finder containerFinder = find.byType(AnimatedContainer);
    Finder opacityFinder = find.byType(AnimatedOpacity);
    Finder centerFinder = find.byType(Center);
    Finder textFinder = find.text(text);

    // Check that the right widgets render the right number of times
    expect(gestureFinder, findsOneWidget);
    expect(opacityFinder, findsOneWidget);
    expect(centerFinder, findsOneWidget);
    expect(textFinder, findsOneWidget);
    expect(containerFinder, findsNWidgets(2));

    // Check that background of default button is the default color
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color,
        equals(Color(0xFF1890FF)));
  });

  testWidgets(
      "Test that custom background colors are used by the widget instead of defaults",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            backgroundColor: Colors.red,
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color, equals(Colors.red));
  });

  testWidgets(
      "Test that disabled buttons use the default color when a custom one is not provided",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            disabled: true,
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color,
        equals(Color(0xFFF5F5F5)));
  });

  testWidgets("Test that key provided is assigned to the root widget",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            key: Key("TestKey"),
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AwesomeButton);
    AwesomeButton root = containerFinder.first.evaluate().single.widget;
    expect(root.key, equals(Key("TestKey")));
  });

  testWidgets("Test that provided font size property is applied",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            fontSize: 26,
          ),
        ),
      ),
    );

    Finder textFinder = find.text(text);
    Text textWidget = textFinder.first.evaluate().single.widget;
    expect(textWidget.style.fontSize, equals(26));
  });

  testWidgets("Test that provided font color property is applied",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            textColor: Colors.red,
          ),
        ),
      ),
    );

    Finder textFinder = find.text(text);
    Text textWidget = textFinder.first.evaluate().single.widget;
    expect(textWidget.style.color, equals(Colors.red));
  });

  testWidgets("Test that provided padding property is applied", (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            padding: EdgeInsets.all(6),
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect(container.padding, equals(EdgeInsets.all(6)));
  });

  testWidgets("Test that provided disabled color property is applied",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            disabled: true,
            disabledColor: Colors.red,
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color, equals(Colors.red));
  });

  testWidgets("Test that provided border radius is applied", (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            borderRadius: 9,
          ),
        ),
      ),
    );

    Finder containerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container = containerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).borderRadius,
        equals(BorderRadius.all(Radius.circular(9))));
  });

  /**
   * Test that the UI changes correctly when tapping it fpr all interact modes
   */
  testWidgets("Test that on pressing the button the background color changes",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color,
        equals(_getActiveColor(Color(0xFF1890FF))));
  });

  testWidgets(
      "Test that given a custom background color, on pressing the button the background color changes accordingly",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            backgroundColor: Colors.purple,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color,
        equals(_getActiveColor(Colors.purple)));
  });

  testWidgets(
      "Test that given a custom active color, on pressing the button the background color changes accordingly",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            activeColor: Colors.purple,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect(
        (container.decoration as BoxDecoration).color, equals(Colors.purple));
  });

  testWidgets(
      "Test that for interact mode opacity, on press the opacity changes accordingly",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            interactMode: InteractMode.OPACITY,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedOpacity);
    AnimatedOpacity container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect(container.opacity, equals(0.3));
  });

  testWidgets(
      "Test that given active opacity for interact mode opacity, on press the opacity changes accordingly",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            interactMode: InteractMode.OPACITY,
            activeOpacity: 0.1,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedOpacity);
    AnimatedOpacity container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect(container.opacity, equals(0.1));
  });

  testWidgets(
      "Test that on pressing the button the background color does not change when disabled",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            disabled: true,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedContainer);
    AnimatedContainer container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect((container.decoration as BoxDecoration).color,
        equals(Color(0xFFF5F5F5)));
  });

  testWidgets(
      "Test that for interact mode opacity, on press the opacity does not change when disabled",
      (tester) async {
    String text = "Test";
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: AwesomeButton(
            text: text,
            interactMode: InteractMode.OPACITY,
            disabled: true,
          ),
        ),
      ),
    );

    await tester.startGesture(tester.getCenter(find.text(text)));
    await tester.pump();

    Finder animatedContainerFinder = find.byType(AnimatedOpacity);
    AnimatedOpacity container =
        animatedContainerFinder.last.evaluate().single.widget;
    expect(container.opacity, equals(1));
  });

  testWidgets("Test that tap down and up are called, and in the correct order",
      (tester) async {
    // The widget uses futures, without runAsync the test will throw a pending timer error
    await tester.runAsync(() async {
      String text = "Test";
      List<String> tapResult = [];
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: AwesomeButton(
              text: text,
              onTapDown: (_) {
                tapResult.add("down");
              },
              onTapUp: (_) {
                tapResult.add("up");
              },
            ),
          ),
        ),
      );

      await tester.tap(find.text(text));
      await tester.pump();

      expect(tapResult.length, equals(2));
      expect(tapResult[0], equals("down"));
      expect(tapResult[1], equals("up"));
    });
  });

  testWidgets(
      "Test that tap down and up are not called when button is disabled",
      (tester) async {
    // The widget uses futures, without runAsync the test will throw a pending timer error
    await tester.runAsync(() async {
      String text = "Test";
      List<String> tapResult = [];
      await tester.pumpWidget(
        MaterialApp(
          home: Material(
            child: AwesomeButton(
              text: text,
              disabled: true,
              onTapDown: (_) {
                tapResult.add("down");
              },
              onTapUp: (_) {
                tapResult.add("up");
              },
            ),
          ),
        ),
      );

      await tester.tap(find.text(text));
      await tester.pump();

      expect(tapResult.length, equals(0));
    });
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'default_sized_buttons.dart';
import 'sized_buttons.dart';
import 'interactions.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      onGenerateRoute: (settings){
        switch(settings.name) {
          case "/":
            return DefaultSized.route();
          case "/sized":
            return Sized.route();
          case "/interact":
            return Interaction.route();
          default:
            return DefaultSized.route();
            break;
        }
      },
      initialRoute: "/",
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[

        ],
      )
    );
  }
}


example/lib/interactions.dart
import 'package:flutter/material.dart';
import 'package:awesome_ui_button/awesome_ui_button.dart';

class Interaction extends StatelessWidget {
  static MaterialPageRoute route() {
    return MaterialPageRoute(builder: (_) {
      return Interaction();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Container(
          padding: EdgeInsets.all(32),
          child: Column(
            children: <Widget>[
              Row(
                children: <Widget>[
                  AwesomeButton(
                    text: "Default",
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.outline(
                      text: "Outline",
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.destructive(
                      text: "Destructive",
                    ),
                  ),
                ],
              ),

              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Row(
                  children: <Widget>[
                    AwesomeButton(
                      text: "Default opacity",
                      interactMode: InteractMode.OPACITY,
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.outline(
                        text: "Outline opacity",
                        interactMode: InteractMode.OPACITY,
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.destructive(
                        text: "Destructive opacity",
                        interactMode: InteractMode.OPACITY,
                      ),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/sized_buttons.dart
import 'package:flutter/material.dart';
import 'package:awesome_ui_button/awesome_ui_button.dart';

class Sized extends StatelessWidget {
  static MaterialPageRoute route() {
    return MaterialPageRoute(builder: (_) {
      return Sized();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Container(
          padding: EdgeInsets.all(32),
          child: Column(
            children: <Widget>[
              Row(
                children: <Widget>[
                  AwesomeButton(
                    text: "Default Small",
                    size: ButtonSize.SMALL,
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.outline(
                      text: "Outline small",
                      size: ButtonSize.SMALL,
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.destructive(
                      text: "Destructive Small",
                      size: ButtonSize.SMALL,
                    ),
                  ),
                ],
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Row(
                  children: <Widget>[
                    AwesomeButton(
                      text: "Default",
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.outline(
                        text: "Outline",
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.destructive(
                        text: "Destructive",
                      ),
                    ),
                  ],
                ),
              ),
              Padding(
                padding: const EdgeInsets.only(top: 20),
                child: Row(
                  children: <Widget>[
                    AwesomeButton(
                      text: "Default Large",
                      size: ButtonSize.LARGE,
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.outline(
                        text: "Outline Large",
                        size: ButtonSize.LARGE,
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.only(left: 12),
                      child: AwesomeButton.destructive(
                        text: "Destructive Large",
                        size: ButtonSize.LARGE,
                      ),
                    ),
                  ],
                ),
              ),
              Padding(
                padding: EdgeInsets.only(top: 20),
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  children: <Widget>[
                    Container(
                      child: AwesomeButton(
                        text: "Default Full Width",
                        size: ButtonSize.FULL_WIDTH,
                      ),
                      width: 500,
                    ),
                  ],
                ),
              ),
              Padding(
                padding: EdgeInsets.only(top: 10),
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  children: <Widget>[
                    Container(
                      child: AwesomeButton.outline(
                        text: "Outline Full Width",
                        size: ButtonSize.FULL_WIDTH,
                      ),
                      width: 500,
                    ),
                  ],
                ),
              ),
              Padding(
                padding: EdgeInsets.only(top: 10),
                child: Row(
                  mainAxisSize: MainAxisSize.max,
                  children: <Widget>[
                    Container(
                      child: AwesomeButton.destructive(
                        text: "Detructive Full Width",
                        size: ButtonSize.FULL_WIDTH,
                      ),
                      width: 500,
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/default_sized_buttons.dart
import 'package:flutter/material.dart';
import 'package:awesome_ui_button/awesome_ui_button.dart';

class DefaultSized extends StatelessWidget {
  static MaterialPageRoute route() {
    return MaterialPageRoute(builder: (_) {
      return DefaultSized();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Container(
          padding: EdgeInsets.all(32),
          child: Column(
            children: <Widget>[
              Row(
                children: <Widget>[
                  AwesomeButton(
                    text: "Default",
                    activeOpacity: 0,
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.outline(
                      text: "Outline",
                      activeOpacity: 0,
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.only(left: 12),
                    child: AwesomeButton.destructive(
                      text: "Destructive",
                      activeOpacity: 0,
                    ),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}


lib/awesome_ui_button.dart
/// This library allows you to add a good looking button without having to write code and manage animations yourself.
///
/// Custom types of this library:
///   [InteractMode] - decides the type of interaction the button uses
///     - DEFAULT: The color of the button changes when the user presses it
///     - OPACITY: The opacity of the button changes when the user presses it
///
///   [ButtonSize] - The size of the button
///     - NORMAL
///     - SMALL
///     - LARGE
///     - FULL_WIDTH: The button fills with the width of its parent
///
/// Sample Usage
///
/// - Using the default button
///     ```
///       AwesomeButton(
///         text: "Press me",
///         onTapUp: (_) {
///           // do something after the user has lifted their finger
///         },
///       )
///     ```
///
/// - Using the button with outline instead of background color
///   ```
///     AwesomeButton.outline(
///       text: "Press me",
///       onTapUp: (_) {
///         // do something after the user has lifted their finger
///       },
///     )
///   ```
///
/// - Using the destructive button: This button renders with a reddish color, commonly used to provide negative options like 'cancel'
///     ```
///       AwesomeButton.destructive(
///         text: "close",
///         onTapUp: (_) {
///           // do something after the user has lifted their finger
///         },
///       )
///
///
/// For full documentation visit https://flutterawesome.dev/docs/components/button
library awesome_ui_button;

import 'package:flutter/material.dart';
import 'dart:math' show max;

enum _ButtonType {
  DEFAULT,
  OUTLINE,
}

/// Type for the kind of interaction the button should use when the user interacts with it.
enum InteractMode {
  DEFAULT,
  OPACITY,
}

/// Type for the size of the button
enum ButtonSize {
  SMALL,
  LARGE,
  NORMAL,
  FULL_WIDTH,
}

// ignore: must_be_immutable
class AwesomeButton extends StatefulWidget {
  _ButtonType _buttonType;

  /// This decides whether or not the button should respond to pointer events. When set to true, the button will ignore touches.
  /// false by default.
  final bool disabled;

  /// The type of interaction the button should use when it receives a pointer event. Depending on the value the button will either
  /// change it's background color when pressed or opacity. [InteractMode.DEFAULT] by default.
  final InteractMode interactMode;

  /// (required) This sets the text of the button.
  final String text;

  /// The key that is set to the root widget. null by default.
  final Key key;

  /// This allows you to set the background color of the button.
  /// NOTE: for [InteractMode.DEFAULT] the background color when the user presses the button changes according to the background color
  /// provided.
  final Color backgroundColor;

  /// Callback that is executed when the button receives when the pointer is lifted from the button i.e the user lifts their finger.
  /// Passes [TapUpDetails] as a parameter
  final GestureTapUpCallback onTapUp;

  /// Callback that is executed when the button receives when the button receives a pointer down event i.e the user presses down on it.
  /// Passes [TapDownDetails] as a parameter
  final GestureTapDownCallback onTapDown;

  /// Sets the size of the button, [ButtonSize] value. [ButtonSize.NORMAL] by default.
  final ButtonSize size;

  /// Sets the color of the text
  final Color textColor;

  /// Sets the size of the text in the button
  final double fontSize;

  /// Allows you to customise the padding of the button
  final EdgeInsets padding;

  /// Allows you to change the color of the button when it is disabled
  final Color disabledColor;

  /// Allows you to customise the value of the radius for the border
  final double borderRadius;

  /// Allows you to change the opacity of the button when the user touches it with [InteractMode.OPACITY]
  final double activeOpacity;

  /// Allows you to change the background color of the button when the user touches it with [InteractMode.DEFAULT]
  final Color activeColor;

  bool _isForcedOpacity = false;
  bool _isDestructive = false;

  AwesomeButton({
    this.disabled = false,
    this.interactMode = InteractMode.DEFAULT,
    @required this.text,
    this.key,
    this.backgroundColor,
    this.onTapUp,
    this.onTapDown,
    this.size = ButtonSize.NORMAL,
    this.textColor,
    this.fontSize,
    this.padding,
    this.disabledColor,
    this.borderRadius,
    this.activeOpacity,
    this.activeColor,
  }) : super(key: key) {
    this._buttonType = _ButtonType.DEFAULT;
  }

  AwesomeButton.outline({
    this.disabled = false,
    this.interactMode = InteractMode.OPACITY,
    @required this.text,
    this.key,
    this.backgroundColor,
    this.onTapUp,
    this.onTapDown,
    this.size = ButtonSize.NORMAL,
    this.textColor,
    this.fontSize,
    this.padding,
    this.disabledColor,
    this.borderRadius,
    this.activeOpacity,
    this.activeColor,
  }) : super(key: key) {
    this._buttonType = _ButtonType.OUTLINE;
    this._isForcedOpacity = true;
  }

  AwesomeButton.destructive({
    this.disabled = false,
    this.interactMode = InteractMode.DEFAULT,
    @required this.text,
    this.key,
    this.backgroundColor,
    this.onTapUp,
    this.onTapDown,
    this.size = ButtonSize.NORMAL,
    this.textColor,
    this.fontSize,
    this.padding,
    this.disabledColor,
    this.borderRadius,
    this.activeOpacity,
    this.activeColor,
  }) : super(key: key) {
    this._buttonType = _ButtonType.DEFAULT;
    this._isDestructive = true;
  }

  @override
  State<StatefulWidget> createState() {
    return _AwesomeButtonState();
  }
}

class _AwesomeButtonState extends State<AwesomeButton> {
  // State Vars
  Color _backgroundColor;
  bool _isPressedDown = false;

  // Animation durations
  int borderAnimDuration = 250;
  int mainAnimDuration = 150;
  double currentOpacity;
  double borderWidth;
  double flareBorderOpacity;

  // Size dimensions
  double horizontalPadding;
  double height;
  double borderWidthNormal = 0;
  double fontSize;

  // Opacity values
  double normalOpacity = 1;
  double activeOpacity;
  double borderOpacityNormal = 0;

  // Colors
  Color backgroundColorActive;
  Color backgroundColorNormal;
  Color _textColor;
  Color _mainBorderColor;

  Color _getActiveColor(Color baseColor) {
    int red = max(baseColor.red - 20, 0);
    int green = max(baseColor.green - 40, 0);
    int blue = max(baseColor.blue - 40, 0);

    return Color.fromRGBO(red, green, blue, 1);
  }

  void _setDimensionsBasedOnSize() {
    switch (widget.size) {
      case ButtonSize.SMALL:
        horizontalPadding = 7;
        height = 24;
        fontSize = 12;
        break;
      case ButtonSize.LARGE:
        horizontalPadding = 15;
        height = 40;
        fontSize = 16;
        break;
      case ButtonSize.NORMAL:
        horizontalPadding = 15;
        height = 32;
        fontSize = 14;
        break;
      case ButtonSize.FULL_WIDTH:
        horizontalPadding = 15;
        height = 32;
        fontSize = 14;
        break;
    }

    fontSize = widget.fontSize ?? fontSize;
  }

  void _initVars() {
    _setDimensionsBasedOnSize();

    Color previousBgNormal = backgroundColorNormal;
    backgroundColorNormal = widget.backgroundColor ??
        (widget._isDestructive ? Color(0xFFFF4D4F) : Color(0xFF1890FF));
    backgroundColorActive =
        widget.activeColor ?? _getActiveColor(backgroundColorNormal);

    if (backgroundColorNormal != previousBgNormal) {
      _backgroundColor = backgroundColorNormal;
      _mainBorderColor = _backgroundColor;
    }

    _textColor = widget._buttonType == _ButtonType.DEFAULT
        ? Colors.white
        : _backgroundColor;
    _textColor = widget.textColor ?? _textColor;

    activeOpacity = widget.activeOpacity ?? 0.3;

    if (widget.disabled) {
      _backgroundColor = widget.disabledColor ?? Color(0xFFF5F5F5);
      _textColor = Colors.black.withOpacity(0.25);
      _mainBorderColor = widget._buttonType == _ButtonType.OUTLINE
          ? widget.disabledColor ?? Color(0xFFD9D9D9)
          : Color(0xFFD9D9D9);
      currentOpacity = normalOpacity;
      flareBorderOpacity = borderOpacityNormal;
    } else if (!_isPressedDown) {
      _backgroundColor = backgroundColorNormal;
      _mainBorderColor = _backgroundColor;
      currentOpacity = normalOpacity;
      flareBorderOpacity = borderOpacityNormal;
    }

    borderWidth = borderWidthNormal;
  }

  Widget _renderButtonText(BuildContext context) {
    return Text(
      widget.text,
      style: TextStyle(
        fontSize: fontSize,
        color: _textColor,
      ),
    );
  }

  EdgeInsets _getButtonPadding() {
    if (widget.padding != null) {
      return widget.padding;
    }

    return EdgeInsets.only(
      left: horizontalPadding,
      right: horizontalPadding,
      top: 6,
      bottom: 6,
    );
  }

  BoxDecoration _getButtonDecoration() {
    return BoxDecoration(
      color: widget._buttonType == _ButtonType.DEFAULT
          ? _backgroundColor
          : Colors.transparent,
      borderRadius: BorderRadius.all(Radius.circular(widget.borderRadius ?? 4)),
      border: Border.all(
        width: widget._buttonType == _ButtonType.DEFAULT ? 0 : 1,
        color: _mainBorderColor,
      ),
    );
  }

  Widget _renderButton(BuildContext context) {
    return AnimatedContainer(
      duration: Duration(milliseconds: borderAnimDuration),
      decoration: BoxDecoration(
        borderRadius:
            BorderRadius.all(Radius.circular(widget.borderRadius ?? 4)),
        boxShadow: [
          BoxShadow(
            color: _mainBorderColor.withOpacity(flareBorderOpacity),
            spreadRadius: borderWidth,
          ),
        ],
      ),
      child: AnimatedOpacity(
        opacity: currentOpacity,
        duration: Duration(
          milliseconds: mainAnimDuration,
        ),
        child: AnimatedContainer(
          constraints: BoxConstraints(
            minHeight: height,
          ),
          duration: Duration(
            milliseconds: mainAnimDuration,
          ),
          padding: _getButtonPadding(),
          decoration: _getButtonDecoration(),
          child: Center(
            child: _renderButtonText(context),
            widthFactor: widget.size == ButtonSize.FULL_WIDTH ? null : 1,
            heightFactor: 1,
          ),
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    _initVars();
    return GestureDetector(
      child: _renderButton(context),
      onTapDown: (TapDownDetails details) {
        if (widget.disabled) {
          return;
        }
        setState(() {
          _isPressedDown = true;
          if (widget.interactMode == InteractMode.DEFAULT &&
              !widget._isForcedOpacity) {
            _backgroundColor = backgroundColorActive;
          } else {
            currentOpacity = activeOpacity;
          }
        });
        if (widget.onTapDown != null) {
          widget.onTapDown(details);
        }
      },
      onTapUp: (TapUpDetails details) {
        if (widget.disabled) {
          return;
        }
        _isPressedDown = false;
        _onTapUp();

        if (widget.onTapUp != null) {
          widget.onTapUp(details);
        }
      },
    );
  }

  void _onTapUp() async {
    setState(() {
      if (widget.interactMode == InteractMode.DEFAULT &&
          !widget._isForcedOpacity) {
        _backgroundColor = backgroundColorNormal;
      } else {
        currentOpacity = normalOpacity;
      }

      if (widget.size != ButtonSize.FULL_WIDTH) {
        flareBorderOpacity = 0.45;
        borderWidth = 10;
      }
    });

    if (widget.size != ButtonSize.FULL_WIDTH) {
      await Future.delayed(
          Duration(milliseconds: (borderAnimDuration / 3).round()));

      setState(() {
        flareBorderOpacity = 0;
      });

      await Future.delayed(
          Duration(milliseconds: (borderAnimDuration * (2 / 3)).round()));

      setState(() {
        borderWidth = borderWidthNormal;
      });

      await Future.delayed(Duration(milliseconds: borderAnimDuration));

      setState(() {
        flareBorderOpacity = borderOpacityNormal;
      });
    }
  }
}
